pkg/grpc_middleware/dialer/builder.go

```diff
@@ -0,0 +1,319 @@
+func (b *Builder) WithClientTransportCredentials(certs ...tls.Certificate) {
+	b.tlsConfig.Certificates = certs
+}
+
+// GetClientTransportCredentials ...
+func (b *Builder) GetClientTransportCredentials() []tls.Certificate {
+	return b.tlsConfig.Certificates
+}
+
+// WithClientCredentials builds transport credentials for a gRPC client
+func (b *Builder) WithClientCredentials(c credentials.PerRPCCredentials) {
+	b.credentials = c
+}
+
+// GetClientCredentials ...
+func (b *Builder) GetClientCredentials() credentials.PerRPCCredentials {
+	return b.credentials
+}
+
+// SetConnInfo allows passing in the dns and port for the connection, providing
+// flexibility if a consumer wants to set a default and or override
+func (b *Builder) SetConnInfo(dns, port string, tls bool) error {
+	u, err := url.Parse(dns)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("invalid dns: %s", dns))
+	}
+	if u.Scheme != "" {
+		// This will cause the connection to fail silently with a cryptic "context.Deadline exceeded" so we validate for it here.
+		return errors.New(fmt.Sprintf("grpc connection dns must not contain a scheme/protocol: %s", dns))
+	}
+	b.dns = &dns
+
+	i, err := strconv.ParseUint(port, 10, 32)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("invalid port number: %s", port))
+	}
+	if i > 65535 {
+		return errors.Errorf("invalid port number: %s", port)
+	}
+	b.port = &port
+	b.withTLS = &tls
+
+	return nil
+}
+
+// GetConnInfo returns the dns and port that were set, and errors if either
+// were null, allowing verfication prior to attempting the connection
+func (b *Builder) GetConnInfo() (dns string, port string, useTLS bool, err error) {
+	if b.dns == nil || b.port == nil || b.withTLS == nil {
+		return "", "", false, errors.New("Connection info not fully set")
+	}
+	return *b.dns, *b.port, *b.withTLS, nil
+}
+
+// SetConnAddr sets connection information as well as configures some common tls and authentication options
+// based on a provided connection string.
+// the format is tls://hostname:port (or tcp://hostname:port if developing locally without TLS)
+// Optional TLS parameters:
+//   skip_verify=true        ignore server CA verification.
+//   ca_file=./filename.pem  CA of service for verification.
+func (b *Builder) SetConnAddr(addr string) error {
+	cfg, err := ReadConfig(addr)
+	if err != nil {
+		return err
+	}
+	cfg.ApplyToBuilder(b)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Dial returns the client connection to the server.
+// context is ignored unless builder is set to block using WithBlock(true)
+func (b *Builder) Dial(ctx context.Context) (*grpc.ClientConn, error) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	dns, port, withTLS, err := b.GetConnInfo()
+	if err != nil {
+		return nil, fmt.Errorf("target connection parameter missing: dns and/or port not set")
+	}
+
+	addr := dns + ":" + port
+	log.Printf("Target to connect = %s, tls = %t", addr, withTLS)
```

Comment:
Should we diable logging by default and add a param to turn it on?

Reply:
require passing it in

Reply:
Is this needed outside local development? To have a log of the connect attempt? Perhaps it can be removed and the error returned on fail would be of more benefit.